src/MonitorProfiler/Logging/Logger.h

```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma once
+
+#include <string>
+#include "corhlpr.h"
+
+/// <summary>
+/// Logging severity levels.
+/// </summary>
+enum class LogLevel
+{
+    Trace,
+    Debug,
+    Information,
+    Warning,
+    Error,
+    Critical,
+    None
+};
+
+/// <summary>
+/// Interface for logging messages.
+/// </summary>
+DECLARE_INTERFACE(ILogger)
+{
+    /// <summary>
+    /// Writes a log message.
+    /// </summary>
+    STDMETHOD(Log)(LogLevel level, const std::string format, va_list args) PURE;
+
+    inline STDMETHODIMP Log(LogLevel level, const std::string format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        HRESULT hr = Log(level, format, args);
+        va_end(args);
+        return hr;
+    }
+};
+
+// Checks if EXPR is a failed HRESULT
+// If failed, logs the failure and returns the HRESULT
+#define IfFailLogRet_(pLogger, EXPR) \
+    do { \
+        hr = (EXPR); \
+        if(FAILED(hr)) { \
+            if (nullptr != pLogger) { \
+                IfFailRet(pLogger->Log( \
```

Comment:
I'm not sure the Log failure should be overriding the EXPR failure. That may be misleading to callers.

Reply:
Maybe we need some cross plat debug assertion notion. Maybe something that writes out to stderr for debug builds.